src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexCharClass.cs

```diff
@@ -1450,44 +1455,32 @@ internal static unsafe string CharsToStringClass(ReadOnlySpan<char> chars)
 
-            if (chars.Length == 2)
-            {
-                switch (chars[0], chars[1])
-                {
-                    case ('A', 'a'): case ('a', 'A'): return "\0\x0004\0ABab";
-                    case ('B', 'b'): case ('b', 'B'): return "\0\x0004\0BCbc";
-                    case ('C', 'c'): case ('c', 'C'): return "\0\x0004\0CDcd";
-                    case ('D', 'd'): case ('d', 'D'): return "\0\x0004\0DEde";
-                    case ('E', 'e'): case ('e', 'E'): return "\0\x0004\0EFef";
-                    case ('F', 'f'): case ('f', 'F'): return "\0\x0004\0FGfg";
-                    case ('G', 'g'): case ('g', 'G'): return "\0\x0004\0GHgh";
-                    case ('H', 'h'): case ('h', 'H'): return "\0\x0004\0HIhi";
-                    // 'I' and 'i' are missing since depending on the cultuure they may
-                    // have additional mappings.
-                    case ('J', 'j'): case ('j', 'J'): return "\0\x0004\0JKjk";
-                    // 'K' and 'k' are missing since their mapping also includes Kelvin K.
-                    case ('L', 'l'): case ('l', 'L'): return "\0\x0004\0LMlm";
-                    case ('M', 'm'): case ('m', 'M'): return "\0\x0004\0MNmn";
-                    case ('N', 'n'): case ('n', 'N'): return "\0\x0004\0NOno";
-                    case ('O', 'o'): case ('o', 'O'): return "\0\x0004\0OPop";
-                    case ('P', 'p'): case ('p', 'P'): return "\0\x0004\0PQpq";
-                    case ('Q', 'q'): case ('q', 'Q'): return "\0\x0004\0QRqr";
-                    case ('R', 'r'): case ('r', 'R'): return "\0\x0004\0RSrs";
-                    case ('S', 's'): case ('s', 'S'): return "\0\x0004\0STst";
-                    case ('T', 't'): case ('t', 'T'): return "\0\x0004\0TUtu";
-                    case ('U', 'u'): case ('u', 'U'): return "\0\x0004\0UVuv";
-                    case ('V', 'v'): case ('v', 'V'): return "\0\x0004\0VWvw";
-                    case ('W', 'w'): case ('w', 'W'): return "\0\x0004\0WXwx";
-                    case ('X', 'x'): case ('x', 'X'): return "\0\x0004\0XYxy";
-                    case ('Y', 'y'): case ('y', 'Y'): return "\0\x0004\0YZyz";
-                    case ('Z', 'z'): case ('z', 'Z'): return "\0\x0004\0Z[z{";
-                }
+                case 1:
+                    // Special-case ASCII characters to avoid the computation/allocation in this very common case.
+                    if (chars[0] < 128)
+                    {
+                        string[] asciiStrings = s_asciiStrings;
+                        if (chars[0] < asciiStrings.Length)
+                        {
+                            return asciiStrings[chars[0]] ??= $"\0\u0002\0{chars[0]}{(char)(chars[0] + 1)}";
+                        }
+                    }
+                    break;
+
+                case 2:
+                    // Special-case cased ASCII letter pairs to avoid the computation/allocation in this very common case.
```

Comment:
Will this work for `i` and `k`? I suppose so since AFAIK in case insensitive case the chars.Length will never be of size 2 as they include extra mappings but it would be good to double check

Reply:
Also, as a very small NIT, it would be good if we mention in the comment that this case is particularly not only looking for ASCII letter pairs, but ASCII letter pairs that only differ in casing like `A` and `a`.

Reply:
> Will this work for i and k?

Initially I had excluded them, just as they were excluded in the previous hardcoded list.  But then I realized this code isn't deciding what the set should be, that's already been computed... all this is doing is translating that set into the corresponding description.  So it's valid for i and k to be included here, and would work fine if [Ii] and [Kk] are what the case folding produced or, more likely, if the developer explicitly wrote that set.

Reply:
> not only looking for ASCII letter pairs, but ASCII letter pairs that only differ in casing

It says "cased ASCII letter pairs" :)

Reply:
oh, totally missed that, sorry.

Reply:
:smile: